veps/sig-compute/hypervisor-feature-discovery.md

```diff
@@ -0,0 +1,372 @@
+- Test infrastructure maintainers ensuring tests run only where features are supported.
+- End users receiving clear error messages when requesting unsupported features.
+
+## User Stories
+
+1. As a contributor adding a new hypervisor, I can declare which features need validation or should be blocked, and have validation work automatically without scattered conditional logic.
+2. As a test maintainer, I can decorate tests with capability requirements for features that need platform-specific validation and have them automatically skip on unsupported platforms.
+3. As an end user, I receive clear error messages with documentation links when I request features that are explicitly unsupported on my platform.
+4. As a platform engineer, I can declare constraints (e.g., "VGA unsupported on ARM64") once and have validation, test filtering, and error messages handled automatically.
+
+## Repos
+
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+- [kubevirt/enhancements](https://github.com/kubevirt/enhancements) (this VEP)
+
+## Design
+
+### Core Types
+
+```go
+// Core types in pkg/capabilities/types.go
+type CapabilityKey string  // e.g., "graphics.vga", "firmware.secureboot.uefi"
+type SupportLevel int
+
+const (
+    Unregistered SupportLevel = iota  // Not registered (default zero value)
+    Unsupported                  // Explicitly blocked on this platform
+    Experimental                 // Requires feature gate
+    Deprecated                   // Supported but discouraged
+)
+
+type Capability struct {
+    Level       SupportLevel
+    Message     string   // User-facing explanation
+    Since       string   // Version when constraint registered
+    DocLink     string   // Link to documentation
+    GatedBy     string   // Optional: feature gate name
+}
+```
+
+### Capability Definitions
+
+All capabilities are defined once as typed constants in `pkg/capabilities/definitions.go`:
+
+```go
+// pkg/capabilities/definitions.go
+type CapabilityKey string
+
+// Capability constants - each represents a feature that may need validation or blocking
+const (
+    CapGraphicsVGA       CapabilityKey = "graphics.vga"
+    CapGraphicsVirtIO    CapabilityKey = "graphics.virtio"
+    CapSecureBootUEFI    CapabilityKey = "firmware.secureboot.uefi"
+    CapCPUHotplug        CapabilityKey = "cpu.hotplug"
+    // ... all capabilities declared as constants
+)
+```
+
+### Registration Pattern
+
+Capabilities are fully defined when hypervisors register their constraints via a builder pattern. The `Capability` struct populated by the builder contains all necessary metadata (level, message, since, docLink, gatedBy).
+
+Hypervisor+architecture combinations register their constraints using a fluent builder API:
+
+```go
+// pkg/capabilities/registry.go - Matrix registration with builder pattern
+type Capabilities struct {
+    matrix map[string]map[CapabilityKey]Capability // "hypervisor/arch" or "hypervisor" -> capabilities
```

Comment:
This structure hints about the intended usage, i.e. that the users of this data will first query per hypervisor and arch, then look at the capabilities.

First, can we have the usage intention first so we can evaluate the data structure?

An example of a discussion in this regard could be this: We may expect only the supported capabilities to be registered, per whatever the cluster provides. What is not supported, is not registered at initial initialization (whatever the component may be).
Was that considered?